apollo-router/src/plugins/response_cache/plugin.rs

```diff
@@ -1620,58 +1620,102 @@ fn get_invalidation_entity_keys_from_schema(
             })?;
-    let cache_keys = field_def
+    let implements_cache_keys = type_def
+        .implements_interfaces
+        .iter()
+        .filter_map(|interface_comp| {
+            let interface = supergraph_schema.get_interface(interface_comp.name.as_str())?;
+            let cache_keys =
+                interface
+                    .directives
+                    .get_all("join__directive")
+                    .filter_map(|directive| {
+                        get_format_from_cache_tag_directive(
+                            supergraph_schema,
+                            subgraph_name,
+                            subgraph_enums,
+                            directive,
+                        )
+                    });
+
+            Some(cache_keys)
+        })
+        .flatten();
+    let cache_keys = type_def
         .directives
         .get_all("join__directive")
-        .filter_map(|dir| {
-            let name = dir.argument_by_name("name", supergraph_schema).ok()?;
-            if name.as_str()? != CACHE_TAG_DIRECTIVE_NAME {
-                return None;
-            }
-            let is_current_subgraph = dir
-                .argument_by_name("graphs", supergraph_schema)
-                .ok()
-                .and_then(|f| {
-                    Some(
-                        f.as_list()?
-                            .iter()
-                            .filter_map(|graph| graph.as_enum())
-                            .any(|g| {
-                                subgraph_enums.get(g.as_str()).map(|s| s.as_str())
-                                    == Some(subgraph_name)
-                            }),
-                    )
-                })
-                .unwrap_or_default();
-            if !is_current_subgraph {
-                return None;
-            }
-            dir.argument_by_name("args", supergraph_schema)
-                .ok()?
-                .as_object()?
-                .iter()
-                .find_map(|(field_name, value)| {
-                    if field_name.as_str() == "format" {
-                        value.as_str()?.parse::<StringTemplate>().ok()
-                    } else {
-                        None
-                    }
-                })
+        .filter_map(|directive| {
+            get_format_from_cache_tag_directive(
+                supergraph_schema,
+                subgraph_name,
+                subgraph_enums,
+                directive,
+            )
         });
     let mut vars = IndexMap::default();
-    vars.insert("$key".to_string(), Value::Object(entity_keys.clone()));
-    let invalidation_cache_keys = cache_keys
+    let mut key_vars = entity_keys.clone();
+    key_vars.insert(
+        ByteString::from(TYPENAME.to_string()),
+        Value::String(typename.to_string().into()),
+    );
+    vars.insert("$key".to_string(), Value::Object(key_vars.clone()));
```

Comment:
For initial GA release, I propose not to support `$key.__typename` in format values, since `__typename` can be too dynamic for interface cache tags.

We may allow object types to have additional cache tags (in addition to the inherited cache tags from their interfaces) that are more specific for the types, which I think safer than `__typename`.